janus_core/src/retries.rs

```diff
@@ -0,0 +1,315 @@
+///
+///   - a timeout
+///   - a problem establishing a connection
+///   - an HTTP status code indicating a server error
+///
+/// If the request eventually succeeds, the value returned by `request_fn` is returned. If an
+/// unretryable failure occurs or enough transient failures occur, the error returned by the last
+/// call to `f` is returned. Retryable failures are logged.
+///
+/// # TODOs:
+///
+/// This function could take a list of HTTP status codes that should be considered retryable, so
+/// that a caller could opt to retry when it sees 408 Request Timeout or 429 Too Many Requests, but
+/// since none of the servers this is currently used to communicate with ever return those statuses,
+/// we don't yet need that feature.
+pub async fn retry_http_request<Request, ResultFuture>(
+    backoff: ExponentialBackoff,
+    request_fn: Request,
+) -> Result<reqwest::Response, reqwest::Error>
+where
+    Request: Clone + FnMut() -> ResultFuture,
+    ResultFuture: Future<Output = Result<reqwest::Response, reqwest::Error>>,
+{
+    let result = retry(backoff, || async {
+        let mut request_fn = request_fn.clone();
+        let result = request_fn().await;
+
+        // In all branches in this match, we wrap the reqwest::Response or reqwest::Error up in a
+        // Result<reqwest::Response, backoff::Error<Result<reqwest::Response, reqwest::Error>>>>,
+        // which allows us to retry on certain HTTP status codes without discarding the
+        // reqwest::Response, which the caller may need in order to examine its body or headers.
+        match result {
+            Ok(response) => {
+                if response.status().is_server_error()
+                    && response.status() != StatusCode::NOT_IMPLEMENTED
+                {
+                    warn!(?response, "encountered retryable server error");
+                    return Err(backoff::Error::transient(Ok(response)));
+                }
+
+                Ok(response)
+            }
+            Err(error) => {
+                if error.is_timeout() || error.is_connect() {
+                    warn!(?error, "encountered retryable error");
+                    return Err(backoff::Error::transient(Err(error)));
+                }
+
+                if let Some(io_error) = find_io_error(&error) {
+                    if let std::io::ErrorKind::ConnectionRefused
+                    | std::io::ErrorKind::ConnectionReset
+                    | std::io::ErrorKind::ConnectionAborted = io_error.kind()
+                    {
+                        warn!(?error, "encountered retryable error");
+                        return Err(backoff::Error::transient(Err(error)));
+                    }
+                }
+
+                debug!("encountered non-retryable error");
+                Err(backoff::Error::permanent(Err(error)))
+            }
+        }
+    })
+    .await;
+
+    match result {
+        Ok(response) => Ok(response),
+        Err(result) => result,
```

Comment:
Coalescing the two different `Ok` variants here makes me worry about the retry function and its caller agreeing on the semantics of success/error. This concern is more applicable to future support for configuring which status codes result in retries.

One idea, for argument's sake: If the retry function returned a two-variant error enum inside `Err()`, with one variant for `reqwest::Error`, and the other for `reqwest::Response` with a failing status, then we could do all our error categorization "inside" the retry handler, and avoid redundant checks on the result of the retry method. (This would also bring a fringe benefit of making it impossible to forget to check the status.)

Reply:
I had a similar thought, but a two-variant error enum where one variant is `reqwest::Error` and the other  is `reqwest::Response` is exactly `Result<reqwest::Response, reqwest::Error>`, which is what I'm sticking inside the error, so I don't think this would make a difference.

I don't see what error categorization is happening outside of the call to `backoff::retry`. All we do with the result from `retry` is flatten the nested `Err()` into the type expected by the caller of `retry_http_request`.

This can be simplified (arguably) by calling `Result::or_else` instead of using a match to flatten the error case.

Reply:
+1, IMO we should allow the caller to determine whether the retry mechanism considered the request a failure or not.

The suggestion to introduce an error enum would work. We could also use `Result<reqwest::Response, reqwest::Error>` as this enum; this is very close to what is returned by the `retry` call & would make the final return value of this function `Result<reqwest::Response, Result<reqwest::Response, reqwest::Error>>`.

Reply:
Sorry for the ambiguity -- yeah, the caller of `retry_http_request()` currently has to determine whether the `Response` represents a success case or failure case.

Reply:
> other is reqwest::Response is exactly Result<reqwest::Response, reqwest::Error>, which is what I'm sticking inside the error, so I don't think this would make a difference.

The return type is currently `Result<reqwest::Response, reqwest::Error>`. A successful-from-reqwest's-perspective request that leads to too many transient retries will be returned as `Ok(response)`, giving the caller no indication that the success response they're getting is actually a failure as far as the retry machinery is concerned.

I think we should modify this function's return type to `Result<reqwest::Response, Result<reqwest::Response, reqwest::Error>>`. `Ok(response)` would be returned only if the response was not a retry; `Err(Ok(response))` would indicate that this response was considered a failure due to retry issues even though reqwest viewed it as a successful HTTP transaction; and `Err(Err(err))` would indicate some HTTP-or-lower-level error reported by reqwest.

Reply:
I gave this a try in 788e7ae957dc5fbdb9614cb24de8184d2d93fbde and I'm not convinced it's better. Nothing that currently uses `retry_http_request` cares about the distinction between `Ok(response)` and `Err(Ok(response))`, so the upshot is that we now force every caller of `retry_http_request` to deal with that ambiguity, whereas otherwise they can just use `?` and then examine the HTTP status code in `Ok(response)`.

If you both think this approach is better, I'm happy to go with the majority decision, though.

Reply:
IMO, the new approach isn't ambiguous, the original approach is: it makes it ambiguous whether an `Ok(response)` return value indicates (a) success or (b) a transient failure that ran out of time to retry. As David pointed out, this forces every caller to implement (matching!) error-detection logic.

If nested `Result`s are too ugly / too hard to deal with, we could map transient failures to an arbitrary error enum indicating the reason for failure. This is easier to consume, but will lose some fidelity if there's ever a caller that wants to inspect e.g. the response even if it is a failure.

Reply:
Well, the good news is, we can change `retry_http_request` in lockstep with all its users in the future, so we don't have to get it perfectly right today. I'll merge this with your preferred approach.